### ISSUE TITLE
[SOLVED] ip addr add dev tun0 10.8.0.2/24 broadcast 10.8.0.255 RTNETLINK answers: File exists

### ISSUE THREAD

**Issue by mixpc** [pivpn/pivpn#496]
Sorry, I have run into the same issue as [closed issue](https://github.com/pivpn/pivpn/issues/269) but here DNS does not seem to work.

Pi vpn client connects to Pi vpn server but DNS does not seem to work. However, mobile and laptop vpn clients connect to Pi vpn server without DNS issues and so can surf the web etc. Please, any advice is welcome.

The same error message is displayed:

`ip addr add dev tun0 10.8.0.2/24 broadcast 10.8.0.255 RTNETLINK answers: File exists`

I have created certificates for three different clients in three different networks. Two of them (laptop and mobile) access Pi VPN server successfully, no DNS issues and can surf the web. Router por forwarding correctly set.

When trying to access Pi VPN server from third client (another Pi, vpn client), the message above is displayed. Many times 10.8.0.2 is assigned to either laptop or mobile. But that address is not busy when trying to use Pi as VPN client.

```
pi@vpnclient:~ $ cat /etc/openvpn/client.conf
client
dev tun
proto udp
remote pi.vpnserver.org 1234
resolv-retry infinite
nobind
persist-key
persist-tun
key-direction 1
remote-cert-tls server
tls-version-min 1.2
verify-x509-name server_Di88e4h6gi9erWQU name
cipher AES-256-CBC
auth SHA256
comp-lzo
verb 1
<ca>
</ca>
<cert>
</cert>
<key>
</key>
<tls-auth>
</tls-auth>

```

```
pi@vpnserver:~ $ sudo nano /etc/openvpn/server.conf
dev tun
proto udp
port 1234
ca /etc/openvpn/easy-rsa/pki/ca.crt
cert /etc/openvpn/easy-rsa/pki/issued/server_Di88e4h6gi9erWQU.crt
key /etc/openvpn/easy-rsa/pki/private/server_Di88e4h6gi9erWQU.key
dh /etc/openvpn/easy-rsa/pki/dh1024.pem
topology subnet
server 10.8.0.0 255.255.255.0
# server and remote endpoints
ifconfig 10.8.0.1 10.8.0.2
# Add route to Client routing table for the OpenVPN Server
push "route 10.8.0.1 255.255.255.255"
# Add route to Client routing table for the OPenVPN Subnet
push "route 10.8.0.0 255.255.255.0"
# your local subnet
push "route 0.0.0.0 "
# Set your primary domain name server address for clients
push "dhcp-option DNS 192.168.1.4"
# 128.0.0.0/1 rather than 0.0.0.0/0. This has the benefit of
# overriding but not wiping out the original default gateway.
push "redirect-gateway def1"
client-to-client
keepalive 10 120
remote-cert-tls client
tls-version-min 1.2
tls-auth /etc/openvpn/easy-rsa/pki/ta.key 0
cipher AES-256-CBC
auth SHA256
comp-lzo
user nobody
group nogroup
persist-key
persist-tun
crl-verify /etc/openvpn/crl.pem
status /var/log/openvpn-status.log 20
status-version 3
log /var/log/openvpn.log
verb 1
# Generated for use by PiVPN.io

```

However when Pi vpn client connects to Pi vpn server...,

```
pi@vpnclient:~ $ curl ifconfig.co
curl: (6) Could not resolve host: ifconfig.co

```
So even though Pi vpn client is connected as client to vpn server, DNS does not seem to work as expected.

Likewise,

```
pi@vpnclient:~ $ ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=54 time=68.7 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=54 time=69.4 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=54 time=67.4 ms
^C
--- 8.8.8.8 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2002ms
rtt min/avg/max/mdev = 67.423/68.518/69.412/0.851 ms

```
But...

```
pi@vpnclient:~ $ ping www.google.com
ping: www.google.com: Temporary failure in name resolution

```

```
pi@vpnclient:~ $ lsb_release -a
No LSB modules are available.
Distributor ID: Raspbian
Description:    Raspbian GNU/Linux 9.3 (stretch)
Release:        9.3
Codename:       stretch

```

```
pi@vpnclient:~ $ ping 10.8.0.1
PING 10.8.0.1 (10.8.0.1) 56(84) bytes of data.
64 bytes from 10.8.0.1: icmp_seq=1 ttl=64 time=44.4 ms
64 bytes from 10.8.0.1: icmp_seq=2 ttl=64 time=57.4 ms
64 bytes from 10.8.0.1: icmp_seq=3 ttl=64 time=48.3 ms
^C
--- 10.8.0.1 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2002ms
rtt min/avg/max/mdev = 44.449/50.097/57.484/5.461 ms

```

```
pi@vpnclient:~ $ ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.5  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 87jd::39dl:98ol:6ed5:2s8j  prefixlen 64  scopeid 0x20<link>
        ether b8:27:ec:c7:2a:43 txqueuelen 1000  (Ethernet)
        RX packets 128957  bytes 9165593 (8.7 MiB)
        RX errors 0  dropped 3155  overruns 0  frame 0
        TX packets 206375  bytes 249570984 (238.0 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1  (Local Loopback)
        RX packets 254  bytes 216220 (211.1 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 254  bytes 216220 (211.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

tun0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 10.8.0.2  netmask 255.255.255.0  destination 10.8.0.2
        inet6 87jd::39dl:98ol:6ed5:2s8j  prefixlen 64  scopeid 0x20<link>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 100  (UNSPEC)
        RX packets 19056  bytes 1316534 (1.2 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 38253  bytes 2640230 (2.5 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0


```

Maybe additional endpoints should be added so that many vpn clients may connect to the vpn server? I am lost at this stage.

```
pi@vpnserver:~ $ sudo nano /etc/openvpn/server.conf
in the configuration
# server and remote endpoints
ifconfig 10.8.0.1 10.8.0.2
```

```
pi@vpnclient:~ $ sudo iptables -S
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
```

```
pi@vpnclient:~ $ sudo iptables -t nat -S
-P PREROUTING ACCEPT
-P INPUT ACCEPT
-P OUTPUT ACCEPT
-P POSTROUTING ACCEPT
```

What is a fact is that laptop and mobile vpn client work ok with no DNS issues.

Thank you for your time and expertise,

Regards,

**Comment by mixpc** [comment]
In case it may be of help, here are more details:

```
pi@vpnclient:~ $ sudo systemctl status openvpn.service
● openvpn.service - OpenVPN service
   Loaded: loaded (/lib/systemd/system/openvpn.service; enabled; vendor preset: enabled)
   Active: active (exited) since Sun 2018-03-11 14:39:01 CET; 5min ago
  Process: 566 ExecStart=/bin/true (code=exited, status=0/SUCCESS)
 Main PID: 566 (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/openvpn.service

Mar 11 14:39:01 vpnclient systemd[1]: Starting OpenVPN service...
Mar 11 14:39:01 vpnclient systemd[1]: Started OpenVPN service.

```

And...

```
pi@vpnclient:~ $ sudo systemctl status openvpn@vpnclient.service
● openvpn@vpnclient.service - OpenVPN connection to vpnclient
   Loaded: loaded (/lib/systemd/system/openvpn@.service; disabled; vendor preset: enabled)
   Active: active (running) since Sun 2018-03-11 14:38:49 CET; 5min ago
     Docs: man:openvpn(8)
           https://community.openvpn.net/openvpn/wiki/Openvpn23ManPage
           https://community.openvpn.net/openvpn/wiki/HOWTO
  Process: 376 ExecStart=/usr/sbin/openvpn --daemon ovpn-vpnclient --status /run/openvpn/vpnclient.status 10 --cd /etc/openvpn --config /etc/openvpn/vpnclient.
 Main PID: 438 (openvpn)
   CGroup: /system.slice/system-openvpn.slice/openvpn@vpnclient.service
           └─438 /usr/sbin/openvpn --daemon ovpn-vpnclient --status /run/openvpn/vpnclient.status 10 --cd /etc/openvpn --config /etc/openvpn/vpnclient.conf --w

Mar 11 14:39:04 vpnclient ovpn-vpnclient[438]: TCP/UDP: Preserving recently used remote address: [AF_INET]37.135.180.24:52205
Mar 11 14:39:04 vpnclient ovpn-vpnclient[438]: UDP link local: (not bound)
Mar 11 14:39:04 vpnclient ovpn-vpnclient[438]: UDP link remote: [AF_INET]37.135.180.24:52205
Mar 11 14:39:06 vpnclient ovpn-vpnclient[438]: [server_Di8n4yqR12ed5WQU] Peer Connection Initiated with [AF_INET]37.135.180.24:52205
Mar 11 14:39:07 vpnclient ovpn-vpnclient[438]: TUN/TAP device tun0 opened
Mar 11 14:39:07 vpnclient ovpn-vpnclient[438]: do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Mar 11 14:39:07 vpnclient ovpn-vpnclient[438]: /sbin/ip link set dev tun0 up mtu 1500
Mar 11 14:39:07 vpnclient ovpn-vpnclient[438]: /sbin/ip addr add dev tun0 10.8.0.2/24 broadcast 10.8.0.255
Mar 11 14:39:07 vpnclient ovpn-vpnclient[438]: ERROR: Linux route add command failed: external program exited with error status: 2
Mar 11 14:39:07 vpnclient ovpn-vpnclient[438]: Initialization Sequence Completed


```

Last three lines seem to indicate that even tough connection is successful there is an issue with additional device and, as stated, no DNS.

**Comment by orazioedoardo** [comment]
Here is the problem (I think): both your pivpn server and pivpn client are on two networks using the same subnet 192.168.1.0/24.

This confuses the pivpn client because when he tries to resolve a domain name using the DNS 192.168.1.4 (pushed by the server) he doesn’t know whether it is a local ip or a remote ip (of the pivpn server network). Most likely for this reason your pivpn client won’t be able to access the rest of the remote LAN as well.

You need to change one of the two subnets to something different (like 192.168.25.0/24 for example) and update the `push "dhcp-option DNS 192.168.1.4"` accordingly.

**Comment by mixpc** [comment]
Hello and thank you very much for your detailed explanation. Well understood.

As I have access to both routers, I have changed the Pi vpn client router subnet to 192.168.2.1 (actually, should be 192.168.2.0/24) so that it is different from the Pi vpn server router subnet (192.168.1.0/24). I have reset both routers and...

```
pi@vpnserver:~ $ hostname -I
192.168.1.4 10.8.0.1

```

```
pi@vpnclient:~ $ hostname -I
192.168.2.5
```
So Pi vpn client and Pi vpn server are in different subnets now.

But...

```
pi@vpnclient:~ $ sudo openvpn --config /etc/openvpn/vpnclient.conf
Sun Mar 11 21:43:31 2018 OpenVPN 2.4.0 arm-unknown-linux-gnueabihf [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Jul 18 2017
Sun Mar 11 21:43:31 2018 library versions: OpenSSL 1.0.2l  25 May 2017, LZO 2.08
Sun Mar 11 21:43:31 2018 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
Sun Mar 11 21:43:31 2018 TCP/UDP: Preserving recently used remote address: [AF_INET]37.135.180.24:52205
Sun Mar 11 21:43:31 2018 UDP link local: (not bound)
Sun Mar 11 21:43:31 2018 UDP link remote: [AF_INET]17.32.80.31:1234
Sun Mar 11 21:43:32 2018 [server_Di88e4h6gi9erWQU] Peer Connection Initiated with [AF_INET]17.32.80.31:1234
Sun Mar 11 21:43:33 2018 TUN/TAP device tun0 opened
Sun Mar 11 21:43:33 2018 do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Sun Mar 11 21:43:33 2018 /sbin/ip link set dev tun0 up mtu 1500
Sun Mar 11 21:43:33 2018 /sbin/ip addr add dev tun0 10.8.0.2/24 broadcast 10.8.0.255
RTNETLINK answers: File exists
Sun Mar 11 21:43:33 2018 ERROR: Linux route add command failed: external program exited with error status: 2
Sun Mar 11 21:43:33 2018 Initialization Sequence Completed

```

So the same issue shows... and ping www.google.com results in "Temporary failure in name resolution" again. 

I have tried once more with the same result but changing verbose to 5 in case the output might be more meaningful. Here is the [pastebin link](https://pastebin.com/ZYyEbxQz).

As I have changed the Pi vpn client router subnet, I have not changed the Pi vpn server router subnet not the line push "dhcp-option DNS 192.168.1.4" in server.conf

Thank you once more for your support,

**Comment by mixpc** [comment]
Just as an addition, here is the laptop log (Windows OS) with successful connection and DNS as well.

```
Sat Mar 10 21:52:06 2018 OpenVPN 2.4.4 x86_64-w64-mingw32 [SSL (OpenSSL)] [LZO] [LZ4] [PKCS11] [AEAD] built on Sep 26 2017
Sat Mar 10 21:52:06 2018 Windows version 6.2 (Windows 8 or greater) 64bit
Sat Mar 10 21:52:06 2018 library versions: OpenSSL 1.0.2l  25 May 2017, LZO 2.10
Enter Management Password:
Sat Mar 10 21:52:09 2018 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
Sat Mar 10 21:52:10 2018 TCP/UDP: Preserving recently used remote address: [AF_INET]17.32.80.31:1234
Sat Mar 10 21:52:10 2018 UDP link local: (not bound)
Sat Mar 10 21:52:10 2018 UDP link remote: [AF_INET]17.32.80.31:1234
Sat Mar 10 21:52:10 2018 [server_Di88e4h6gi9erWQU] Peer Connection Initiated with [AF_INET]17.32.80.31:1234
Sat Mar 10 21:52:12 2018 open_tun
Sat Mar 10 21:52:12 2018 TAP-WIN32 device [Ethernet 4] opened: \\.\Global\{CB9IK7HU-F2E5-42AB-BAS8-79I7YEECACBF}.tap
Sat Mar 10 21:52:12 2018 Set TAP-Windows TUN subnet mode network/local/netmask = 10.8.0.0/10.8.0.3/255.255.255.0 [SUCCEEDED]
Sat Mar 10 21:52:12 2018 Notified TAP-Windows driver to set a DHCP IP/netmask of 10.8.0.3/255.255.255.0 on interface {CB9IK7HU-F2E5-42AB-BAS8-79I7YEECACBF} [DHCP-serv: 10.8.0.254, lease-time: 31536000]
Sat Mar 10 21:52:12 2018 Successful ARP Flush on interface [11] {CB9IK7HU-F2E5-42AB-BAS8-79I7YEECACBF}
Sat Mar 10 21:52:12 2018 do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Sat Mar 10 21:52:17 2018 Initialization Sequence Completed
Sat Mar 10 22:03:33 2018 SIGTERM[hard,] received, process exiting

```

**Comment by orazioedoardo** [comment]
Why is your pastebin link saying that your pivpn client ip is 192.168.101.5? Also I'm seeing `dhcp-option DNS 192.168.100.4` which is yet another ip.

Try this, in your server config change `push "dhcp-option DNS 192.168.1.4"` to `push "dhcp-option DNS 8.8.8.8"`, restart the server with `sudo systemctl restart openvpn`, then try to connect again.

**Comment by mixpc** [comment]
Pi vpnclient connects to Pi vpn server...
`pi@vpnclient:~ $ sudo openvpn --config /etc/openvpn/rpilacasella.conf`

Pi vpnclient is connected to Pi vpn server
```
pi@vpnserver:~ $ pivpn -c

: NOTE : The output below is NOT real-time!
:      : It may be off by a few minutes.

::: Client Status List :::
                                                                Bytes           Bytes
Name                    Remote IP               Virtual IP      Received        Sent            Connected Since
vpnclient            17.32.80.31:42814    10.8.0.2        4.6KiB          4.3KiB          Mar 12 2018 - 11:08:49

```

And then in Pi vpn client
```
pi@vpnclient:~ $ cat /etc/resolv.conf
# Generated by resolvconf
domain home
nameserver 212.166.211.3
nameserver 62.81.16.164
```

I have no clue about those two servers! I have queried and...
212.166.211.3 is hosted by Vodafone
62.81.16.164 is hosted by Vodafone
As Internet Service Provider is Vodafone I assume these are Vodafone DNS?

Some more details:

```
pi@vpnclient:~ $ ls -al /etc/resolv
resolvconf/      resolv.conf      resolv.conf.bak  resolvconf.conf
```

```
pi@vpnclient:~ $ cat /etc/resolv.conf
# Generated by resolvconf
domain home
nameserver 212.166.211.3
nameserver 62.81.16.164
```

```
pi@vpnclient:~ $ cat /etc/resolv.conf.bak
nameserver 8.8.8.8
```

```
pi@vpnclient:~ $ cat /etc/resolvconf.conf
# Configuration for resolvconf(8)
# See resolvconf.conf(5) for details

resolv_conf=/etc/resolv.conf
# If you run a local name server, you should uncomment the below line and
# configure your subscribers configuration files below.
#name_servers=127.0.0.1

# Mirror the Debian package defaults for the below resolvers
# so that resolvconf integrates seemlessly.
dnsmasq_resolv=/var/run/dnsmasq/resolv.conf
pdnsd_conf=/etc/pdnsd.conf
unbound_conf=/var/cache/unbound/resolvconf_resolvers.conf
```
Now, the line
```
# If you run a local name server, you should uncomment the below line and
# configure your subscribers configuration files below.
#name_servers=127.0.0.1
```
has given a hint? As Pi vpn server runs along PiHole (there is a lighttpd server installed but I don't think that should cause interference), which afaik runs as a DNS server as well. So that may be the reason why Pi vpn client is not getting DNS?

Thank you very much for the nth time,

**Comment by orazioedoardo** [comment]
Add those two lines inside the CLIENT config then try again.
`up /etc/openvpn/update-resolv-conf`
`down /etc/openvpn/update-resolv-conf`

By the way if this works you can hole the pihole dns as well, just change the push dns to `10.8.0.1`

**Comment by mixpc** [comment]
Thank you for your continuous feedback. I have tried two approaches:

With your approach...

```
pi@vpnserver:~ $ sudo nano /etc/openvpn/server.conf
[...]
push "dhcp-option DNS 192.168.100.4"
[...]

```
```
pi@vpnclient:~ $ sudo nano /etc/resolvconf.conf
# If you run a local name server, you should uncomment the below line and
# configure your subscribers configuration files below.
#name_servers=127.0.0.1
```

```
pi@vpnclient:~ $ sudo nano /etc/openvpn/vpnclient.conf
[...]
up /etc/openvpn/update-resolv-conf
down /etc/openvpn/update-resolv-conf
[...]
```

```
pi@vpnclient:~ $ sudo openvpn --script-security 2 --config /etc/openvpn/vpnclient.conf
Mon Mar 12 14:00:19 2018 us=56166 ROUTE_GATEWAY 192.168.2.1/255.255.255.0 IFACE=eth0 HWADDR=b8:27:eb:4a:3f:7a
Mon Mar 12 14:00:19 2018 us=57762 TUN/TAP device tun0 opened
Mon Mar 12 14:00:19 2018 us=57896 TUN/TAP TX queue length set to 100
Mon Mar 12 14:00:19 2018 us=57991 do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Mon Mar 12 14:00:19 2018 us=58109 /sbin/ip link set dev tun0 up mtu 1500
Mon Mar 12 14:00:19 2018 us=71403 /sbin/ip addr add dev tun0 10.8.0.2/24 broadcast 10.8.0.255
Mon Mar 12 14:00:19 2018 us=79250 /etc/openvpn/update-resolv-conf tun0 1500 1553 10.8.0.2 255.255.255.0 init
dhcp-option DNS 192.168.1.4
Too few arguments.
Too few arguments.
Mon Mar 12 14:00:19 2018 us=215254 /sbin/ip route add 17.32.80.31/32 via 192.168.101.1
Mon Mar 12 14:00:19 2018 us=219733 /sbin/ip route add 0.0.0.0/1 via 10.8.0.1
Mon Mar 12 14:00:19 2018 us=223469 /sbin/ip route add 128.0.0.0/1 via 10.8.0.1
Mon Mar 12 14:00:19 2018 us=227059 /sbin/ip route add 10.8.0.1/32 via 10.8.0.1
Mon Mar 12 14:00:19 2018 us=230814 /sbin/ip route add 10.8.0.0/24 via 10.8.0.1
RTNETLINK answers: File exists
Mon Mar 12 14:00:19 2018 us=234798 ERROR: Linux route add command failed: external program exited with error status: 2
Mon Mar 12 14:00:19 2018 us=235056 /sbin/ip route add 0.0.0.0/32 via 10.8.0.1
Mon Mar 12 14:00:19 2018 us=238457 Initialization Sequence Completed
```

Open a new session in Pi vpn client and...

```
pi@vpnclient:~ $ ping www.google.com
PING www.google.com (172.217.16.228) 56(84) bytes of data.
64 bytes from mad08s04-in-f4.1e100.net (172.217.16.228): icmp_seq=1 ttl=56 time=62.6 ms
64 bytes from mad08s04-in-f4.1e100.net (172.217.16.228): icmp_seq=2 ttl=56 time=62.7 ms
64 bytes from mad08s04-in-f4.1e100.net (172.217.16.228): icmp_seq=3 ttl=56 time=73.7 ms
64 bytes from mad08s04-in-f4.1e100.net (172.217.16.228): icmp_seq=4 ttl=56 time=61.7 ms
^C
--- www.google.com ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3004ms
rtt min/avg/max/mdev = 61.766/65.226/73.752/4.946 ms
```

So it is successful! Pi vpn client connects successfully and DNS is ok. However, the ERROR message is still displayed.


The second approach I have tried is...

leave push dns as it is
```
pi@vpnserver:~ $ sudo nano /etc/openvpn/server.conf
[...]
push "dhcp-option DNS 192.168.100.4"
[...]
```

uncomment name_servers and edit to 192.168.1.4
```
pi@vpnclient:~ $ sudo nano /etc/resolvconf.conf
# If you run a local name server, you should uncomment the below line and
# configure your subscribers configuration files below.
name_servers=192.168.1.4
```

leave vpnclient.conf as it originally were, not adding
up /etc/openvpn/update-resolv-conf
down /etc/openvpn/update-resolv-conf

And it is successful as well:

```
pi@vpnclient:~ $ ping www.google.com
PING www.google.com (216.58.210.132) 56(84) bytes of data.
64 bytes from mad06s09-in-f132.1e100.net (216.58.210.132): icmp_seq=1 ttl=56 time=61.2 ms
64 bytes from mad06s09-in-f132.1e100.net (216.58.210.132): icmp_seq=2 ttl=56 time=67.4 ms
64 bytes from mad06s09-in-f132.1e100.net (216.58.210.132): icmp_seq=3 ttl=56 time=84.4 ms
^C
--- www.google.com ping statistics ---
4 packets transmitted, 3 received, 25% packet loss, time 3004ms
rtt min/avg/max/mdev = 61.215/71.030/84.425/9.807 ms
```

The connection log shows...

```
pi@vpnclient:~ $ sudo openvpn --config /etc/openvpn/vpnclient.conf
[...log is clipped...]
Mon Mar 12 12:07:28 2018 us=684111 ROUTE_GATEWAY 192.168.2.1/255.255.255.0 IFACE=eth0 HWADDR=b8:27:eb:4a:3f:7a
Mon Mar 12 12:07:28 2018 us=696197 TUN/TAP device tun0 opened
Mon Mar 12 12:07:28 2018 us=696409 TUN/TAP TX queue length set to 100
Mon Mar 12 12:07:28 2018 us=696551 do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Mon Mar 12 12:07:28 2018 us=696704 /sbin/ip link set dev tun0 up mtu 1500
Mon Mar 12 12:07:28 2018 us=710480 /sbin/ip addr add dev tun0 10.8.0.2/24 broadcast 10.8.0.255
Mon Mar 12 12:07:28 2018 us=717785 /sbin/ip route add 17.32.80.31/32 via 192.168.2.1
Mon Mar 12 12:07:28 2018 us=724719 /sbin/ip route add 0.0.0.0/1 via 10.8.0.1
Mon Mar 12 12:07:28 2018 us=731413 /sbin/ip route add 128.0.0.0/1 via 10.8.0.1
Mon Mar 12 12:07:28 2018 us=737880 /sbin/ip route add 10.8.0.1/32 via 10.8.0.1
Mon Mar 12 12:07:28 2018 us=744737 /sbin/ip route add 10.8.0.0/24 via 10.8.0.1
RTNETLINK answers: File exists
Mon Mar 12 12:07:28 2018 us=751720 ERROR: Linux route add command failed: external program exited with error status: 2
Mon Mar 12 12:07:28 2018 us=751955 /sbin/ip route add 0.0.0.0/32 via 10.8.0.1
Mon Mar 12 12:07:28 2018 us=761824 Initialization Sequence Completed
WrWrWR
```

In this second approach Pi vpn client connects successfully and DNS is ok. However, the ERROR message is still displayed as well. In this approach there is not a "Too few arguments" message, which is logical as the config files does not incude
up /etc/openvpn/update-resolv-conf
down /etc/openvpn/update-resolv-conf

So it seems there are actually two issues. One, DNS in Pi vpn client that is now solved (I think your approach is better and mine by tweaking name_servers will not work once vpn connection is closed, as 192.168.1.4 is only available on the vpn server side). And two, the same ERROR that still persists... I cannot fin a reason why the error is still there and woud like to solve it as well to leave a correct vpn connection...

Regards,

**Comment by orazioedoardo** [comment]
Can you just use this server config (i removed redundant routes):

> dev tun
proto udp
port 1234
ca /etc/openvpn/easy-rsa/pki/ca.crt
cert /etc/openvpn/easy-rsa/pki/issued/server_Di88e4h6gi9erWQU.crt
key /etc/openvpn/easy-rsa/pki/private/server_Di88e4h6gi9erWQU.key
dh /etc/openvpn/easy-rsa/pki/dh1024.pem
topology subnet
server 10.8.0.0 255.255.255.0
push "dhcp-option DNS 8.8.8.8"
push "redirect-gateway def1"
client-to-client
keepalive 10 120
remote-cert-tls client
tls-version-min 1.2
tls-auth /etc/openvpn/easy-rsa/pki/ta.key 0
cipher AES-256-CBC
auth SHA256
comp-lzo
user nobody
group nogroup
persist-key
persist-tun
crl-verify /etc/openvpn/crl.pem
status /var/log/openvpn-status.log 20
status-version 3
log /var/log/openvpn.log
verb 1

Restart the server and REBOOT the client, don’t mess with resolv.conf file because they will be overwritten.

**Comment by mixpc** [comment]
Hello,

I can see your point. Remove all push "route and change push dns to Google's DNS. I have done as you suggested and...

```
Mon Mar 12 15:06:25 2018 us=958649 ROUTE_GATEWAY 192.168.2.1/255.255.255.0 IFACE=eth0 HWADDR=b8:27:eb:4a:3f:7a
Mon Mar 12 15:06:25 2018 us=960267 TUN/TAP device tun0 opened
Mon Mar 12 15:06:25 2018 us=960443 TUN/TAP TX queue length set to 100
Mon Mar 12 15:06:25 2018 us=960585 do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Mon Mar 12 15:06:25 2018 us=960760 /sbin/ip link set dev tun0 up mtu 1500
Mon Mar 12 15:06:25 2018 us=969945 /sbin/ip addr add dev tun0 10.8.0.2/24 broadcast 10.8.0.255
Mon Mar 12 15:06:25 2018 us=977700 /etc/openvpn/update-resolv-conf tun0 1500 1553 10.8.0.2 255.255.255.0 init
dhcp-option DNS 8.8.8.8
Too few arguments.
Too few arguments.
Mon Mar 12 15:06:26 2018 us=116172 /sbin/ip route add 17.32.80.31/32 via 192.168.101.1
Mon Mar 12 15:06:26 2018 us=119577 /sbin/ip route add 0.0.0.0/1 via 10.8.0.1
Mon Mar 12 15:06:26 2018 us=122986 /sbin/ip route add 128.0.0.0/1 via 10.8.0.1
Mon Mar 12 15:06:26 2018 us=126427 Initialization Sequence Completed
WrWrWRrWRW
```

And...
```

pi@vpnclient:~ $ ping www.google.com
PING www.google.com (216.58.210.132) 56(84) bytes of data.
64 bytes from mad06s09-in-f132.1e100.net (216.58.210.132): icmp_seq=1 ttl=57 time=173 ms
64 bytes from mad06s09-in-f132.1e100.net (216.58.210.132): icmp_seq=2 ttl=57 time=180 ms
^C
--- www.google.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 173.317/177.118/180.920/3.824 ms

```
So Pi vpn client connection is ok, DNS is ok and ERROR is gone. Success!

To round things up, just a few doubts:

Could push "dhcp-option DNS 8.8.8.8" be changed to push "dhcp-option DNS 192.168.1.4" to take PiHole DNS?
Is there a need for the removed?
push "route 10.8.0.1 255.255.255.255"
push "route 10.8.0.0 255.255.255.0"
push "route 0.0.0.0 "
I can't think much about "Too few argments" as it does not seem to cause any issues.

Thank you very much for the nth+1 time,

**Comment by orazioedoardo** [comment]
To use pihole DNS, go to the pihole web interface -> setting -> DNS -> Interface listening behavior and set the first option (listen on all interfaces), then change the push directive to `push “dhcp-option DNS 10.8.0.1”`.

Important! To make sure that DNS queries are actually going through the tunnel, install tcpdump on the client, then open two terminal windows (still on the client) and run on the first window `sudo tcpdump -i eth0 src port 53`, then `ping google.com` on the second one. You should expect NO traffic output on the first window. Also check if the public ip is indeed the server ip with `wget -qO- http://ipecho.net/plain`.

**Comment by mixpc** [comment]
I have checked  Interface listening behavior and it was set on "Listen only on interface wlan0" (yes, Pi vpn server is connected via WiFi, no eth0 connection possible at the moment because of distance). So I have not changed that setting. I have changed push directive to
push "dhcp-option DNS 10.8.0.1"
After connecting Pi vpn client to Pi vpn server (terminal window #1)
I have entered (terminal window#2)
wget -qO- http://ipecho.net/plain
and the change of ip address has been obvious compared to same command before joining vpn. Good.
And
sudo tcpdump -i eth0 src port 53 (terminal window #3)
has not showed any output when pinging www.google.com (terminal window #2). Thank you very much for your explanation about tcpdump and how to use to monitor. Really useful.

In regards to Too few arguments, I don't think it causes an issue, so I don't think I should be very concerned about it. The command I run is:
`pi@vpnclient:~ $ sudo openvpn --config /etc/openvpn/vpnclient.conf`

```
pi@vpnclient:~ $ sudo cat /etc/openvpn/vpnclient.conf
client
dev tun
proto udp
remote pi.vpnserver.org 1234
resolv-retry infinite
nobind
persist-key
persist-tun
key-direction 1
remote-cert-tls server
tls-version-min 1.2
verify-x509-name server_Di8n4yqR12ed5WQU name
cipher AES-256-CBC
auth SHA256
comp-lzo
verb 5
up /etc/openvpn/update-resolv-conf
down /etc/openvpn/update-resolv-conf
script-security 2

```

Thank you for your observation about 1024-bit. This is a personal configuration (no enterprise whatsoever) so I think I will 4096 anytime soon. By the way, pivpn used to rate 4096 as paranoid when installing. Now, it's not rating it as such anymore.
